### PR TITLE
Optimize compilation of functional-looking assignment RHSs

### LINKED ISSUE #3868: Nested LetExpr chain compiles to tons of nested Java lambdas, which javac chokes on
### Dafny version

4.0.0

### Code to produce this issue

```dafny
include "Wrappers.dfy"

module Test {

  import opened Wrappers

  datatype Or = A | B

  function WoahThatsDeep(o: Or, x: string): Option<string> {
    var r :- match o {
      case A =>
        var x1 := x;
        var x2 := x1;
        var x3 := x2;
        var x4 := x3;
        var x5 := x4;
        var x6 := x5;
        var x7 := x6;
        var x8 := x7;
        var x9 := x8;
        var x10 := x9;
        Some(x10)
      case B =>
        Some("hello")
    };
    Some(r)
  }

  method Main() {
    print WoahThatsDeep(A, "42"), "\n";
  }
}
```

### Command to run and resulting output

```code
dafny run -t:java Scratch.dfy

Takes a very long time to compile the resulting Java code, and uses multiple GB of RAM in the process. This example is simplified from a larger example that causes `javac` to run out of memory.
```

### What happened?

The translated java code looks like this:

```java
// Class __default
// Dafny class __default compiled into Java
package Test_Compile;

import Wrappers_Compile.*;

@SuppressWarnings({"unchecked", "deprecation"})
public class __default {
  public __default() {
  }
  public static Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>> WoahThatsDeep(Or o, dafny.DafnySequence<? extends Character> x)
  {
    Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>> _18_valueOrError0 = ((java.util.function.Function<Or, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>)(_source6_boxed0) -> {
      Or _source6 = ((Or)(java.lang.Object)(_source6_boxed0));
      if (_source6.is_A()) {
        return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(x, boxed0 -> {
          dafny.DafnySequence<? extends Character> _pat_let0_0 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed0));
          return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_pat_let0_0, boxed1 -> {
            dafny.DafnySequence<? extends Character> _19_x1 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed1));
            return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_19_x1, boxed2 -> {
              dafny.DafnySequence<? extends Character> _pat_let1_0 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed2));
              return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_pat_let1_0, boxed3 -> {
                dafny.DafnySequence<? extends Character> _20_x2 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed3));
                return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_20_x2, boxed4 -> {
                  dafny.DafnySequence<? extends Character> _pat_let2_0 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed4));
                  return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_pat_let2_0, boxed5 -> {
                    dafny.DafnySequence<? extends Character> _21_x3 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed5));
                    return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_21_x3, boxed6 -> {
                      dafny.DafnySequence<? extends Character> _pat_let3_0 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed6));
                      return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_pat_let3_0, boxed7 -> {
                        dafny.DafnySequence<? extends Character> _22_x4 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed7));
                        return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_22_x4, boxed8 -> {
                          dafny.DafnySequence<? extends Character> _pat_let4_0 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed8));
                          return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_pat_let4_0, boxed9 -> {
                            dafny.DafnySequence<? extends Character> _23_x5 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed9));
                            return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_23_x5, boxed10 -> {
                              dafny.DafnySequence<? extends Character> _pat_let5_0 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed10));
                              return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_pat_let5_0, boxed11 -> {
                                dafny.DafnySequence<? extends Character> _24_x6 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed11));
                                return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_24_x6, boxed12 -> {
                                  dafny.DafnySequence<? extends Character> _pat_let6_0 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed12));
                                  return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_pat_let6_0, boxed13 -> {
                                    dafny.DafnySequence<? extends Character> _25_x7 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed13));
                                    return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_25_x7, boxed14 -> {
                                      dafny.DafnySequence<? extends Character> _pat_let7_0 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed14));
                                      return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_pat_let7_0, boxed15 -> {
                                        dafny.DafnySequence<? extends Character> _26_x8 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed15));
                                        return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_26_x8, boxed16 -> {
                                          dafny.DafnySequence<? extends Character> _pat_let8_0 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed16));
                                          return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_pat_let8_0, boxed17 -> {
                                            dafny.DafnySequence<? extends Character> _27_x9 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed17));
                                            return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_27_x9, boxed18 -> {
                                              dafny.DafnySequence<? extends Character> _pat_let9_0 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed18));
                                              return ((Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>)(java.lang.Object)(dafny.Helpers.<dafny.DafnySequence<? extends Character>, Wrappers_Compile.Option<dafny.DafnySequence<? extends Character>>>Let(_pat_let9_0, boxed19 -> {
                                                dafny.DafnySequence<? extends Character> _28_x10 = ((dafny.DafnySequence<? extends Character>)(java.lang.Object)(boxed19));
                                                return Wrappers_Compile.Option.<dafny.DafnySequence<? extends Character>>create_Some(_28_x10);
                                              }
                                              )));
                                            }
                                            )));
                                          }
                                          )));
                                        }
                                        )));
                                      }
                                      )));
                                    }
                                    )));
                                  }
                                  )));
                                }
                                )));
                              }
                              )));
                            }
                            )));
                          }
                          )));
                        }
                        )));
                      }
                      )));
                    }
                    )));
                  }
                  )));
                }
                )));
              }
              )));
            }
            )));
          }
          )));
        }
        )));
      } else {
        return Wrappers_Compile.Option.<dafny.DafnySequence<? extends Character>>create_Some(dafny.DafnySequence.asString("hello"));
      }
    }).apply(o);
    if ((_18_valueOrError0).IsFailure(dafny.DafnySequence.<Character>_typeDescriptor(dafny.TypeDescriptor.CHAR))) {
      return (_18_valueOrError0).<dafny.DafnySequence<? extends Character>>PropagateFailure(dafny.DafnySequence.<Character>_typeDescriptor(dafny.TypeDescriptor.CHAR), dafny.DafnySequence.<Character>_typeDescriptor(dafny.TypeDescriptor.CHAR));
    } else {
      dafny.DafnySequence<? extends Character> _29_r = (_18_valueOrError0).Extract(dafny.DafnySequence.<Character>_typeDescriptor(dafny.TypeDescriptor.CHAR));
      return Wrappers_Compile.Option.<dafny.DafnySequence<? extends Character>>create_Some(_29_r);
    }
  }
  public static <__T> Wrappers_Compile.Option<__T> Maybe(dafny.TypeDescriptor<__T> _td___T, __T t)
  {
    return Wrappers_Compile.Option.<__T>create_Some(t);
  }
  public static void Main(dafny.DafnySequence<? extends dafny.DafnySequence<? extends Character>> __noArgsParameter)
  {
    System.out.print(java.lang.String.valueOf(__default.WoahThatsDeep(Test_Compile.Or.create_A(), dafny.DafnySequence.asString("42"))));
    System.out.print((dafny.DafnySequence.asString("\n")).verbatimString());
  }
  public static void __Main(dafny.DafnySequence<? extends dafny.DafnySequence<? extends Character>> args) {
    __default.Main(args);
  }
  private static final dafny.TypeDescriptor<__default> _TYPE = dafny.TypeDescriptor.referenceWithInitializer(__default.class, () -> (__default) null);
  public static dafny.TypeDescriptor<__default> _typeDescriptor() {
    return (dafny.TypeDescriptor<__default>) (dafny.TypeDescriptor<?>) _TYPE;
  }
  @Override
  public java.lang.String toString() {
    return "Test_Compile._default";
  }
}
```

Normally a chain of let expressions (i.e. `var x1 := x; var x2 := x1; ...`) is optimized to compile to a sequence of regular variable declarations in Java, but because this one is inside a match expression rather than at the top level, the optimization isn't currently applied.

### What type of operating system are you experiencing the problem on?

Mac

### LINKED ISSUE #5597: Missing downcasts
### Dafny version

4.7.0

### Code to produce this issue

```dafny
method Main() {
  var n: set<Number> := {};
  var s: set<Integer>;
  s := DoItWithAssignment(n);
  print |s|, " ";
  s := DoItWithPlainLet(n);
  print |s|, " ";
  s := DoItWithOptimizedLet(n);
  print |s|, " ";
  s := DoItViaFunctionBodyResult(n);
  print |s|, "\n";
}

trait Number {
  const value: int
}

class Integer extends Number {
  constructor(value: int) {
    this.value := value;
  }
}

method DoItWithAssignment(numbers: set<Number>) returns (integers00: set<Integer>)
  requires |numbers| == 0
{
  integers00 := numbers;
}

function DoItWithPlainLet(numbers: set<Number>): set<Integer>
  requires |numbers| == 0
{
  {} +
  var integers11: set<Integer> := numbers;
  integers11
}

function DoItWithOptimizedLet(numbers: set<Number>): set<Integer>
  requires |numbers| == 0
{
  var integers22: set<Integer> := numbers;
  integers22
}

function DoItViaFunctionBodyResult(numbers: set<Number>): set<Integer>
  requires |numbers| == 0
{
  numbers
}
```

### Command to run and resulting output

```code
% dafny run test.dfy

Dafny program verifier finished with 5 verified, 0 errors
Errors compiling program into test
source(5770,43): error CS0266: Cannot implicitly convert type 'Dafny.ISet<_module.Number>' to 'Dafny.ISet<_module.Integer>'. An explicit conversion exists (are you missing a cast?)

source(5774,14): error CS0266: Cannot implicitly convert type 'Dafny.ISet<_module.Number>' to 'Dafny.ISet<_module.Integer>'. An explicit conversion exists (are you missing a cast?)
```

### What happened?

Dafny emits malformed code for the bodies of `DoItWithOptimizedLet` and `DoItViaFunctionBodyResult`.

### What type of operating system are you experiencing the problem on?

Mac

### PR DESCRIPTION
**Pull request by RustanLeino**
### Description

Previously, the single-pass compiler would use `TrExprOpt` to turn nested expressions into target-language statements, ending in a `return`. This PR generalizes that mechanism so that the ending action is parameterized. In particular, the PR now applies `TrExprOpt` to single-LHS assignment statements. This allows the RHS of such assignment statements to be deeply nested expressions, without running into, for example, the limits of `javac` to deal with nested lambda expressions.

Fixes #3868

### Implementation

This PR required a change to the target-code strategy of `match` constructs. Previously, `case`s turned into a sequence of `if` statements (note, not a cascading `if`). If one of those `if` statements determined that the `case` applied, then a boolean variable, `unmatched`, was set to `false`. Each of these `if` statements was thus also guarded by the condition `unmatched`.

This previous strategy caused a problem when the body of the cases is an assignment. Since every assignment was guarded by an `if`, target languages with definite-assignment rules weren't able to determine that the variable was always assigned.

A more straightforward strategy is to jump (using a `break` or `goto`) to the end of the target code for the `match`, rather than setting the `unmatched` variable. This lets the target language easily see that the variable has been assigned on all paths. It also means that the `unmatched` variable and its tests are not needed at all. Furthermore, the last `case` does not need an enclosing `if` structure at all, since if the prior `case`s don't apply, then the last one must apply.

This PR improves the strategy to the one with jumps.

### Work along the way

fix: Use Downcast for let RHS and for function body

Fixes #5597

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

**Review by atomb — DISMISSED** [review]
This looks good to me. I'm happy to see that it's a relatively simple change!

**Comment by atomb** [comment]
I see a couple of compilation tests failing, [here](https://github.com/dafny-lang/dafny/actions/runs/9788481880/job/27026634549?pr=5589#step:24:224) and [here](https://github.com/dafny-lang/dafny/actions/runs/9788481880/job/27026634641?pr=5589#step:24:460).

File: Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.cs [hunk 1]
```diff
@@ -3014,57 +3014,8 @@ public record OptimizedExpressionContinuation(Action<Expression, Type, bool, Con
...
-              inTmps[n],
-              EmitAssignment(IdentLvalue(IdName(p)), p.Type, inTypes[n], wr, e.tok)
-            );
-            n++;
-          }
-        }
-        Contract.Assert(n == inTmps.Count);
-        // finally, the jump back to the head of the function
-        EmitJumpToTailCallStart(wr);
+        Contract.Assert(!inLetExprBody); // a tail call had better not sit inside a target-code lambda
```
Comment (atomb) [inline Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.cs:3025]: Nice consolidation!

File: Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.cs [hunk 2]
```diff
@@ -2749,7 +2750,10 @@ protected class NullClassWriter : IClassWriter {
         Coverage.Instrument(f.Body.tok, $"entry to function {f.FullName}", w);
         Contract.Assert(enclosingFunction == null);
         enclosingFunction = f;
+        CompileReturnBody(f.Body, f.OriginalResultTypeWithRenamings(), w, accVar);
+#if NEW_ATTEMPT
```
Comment (atomb) [inline Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.cs:2754]: Is this relevant beyond debugging?

File: Source/DafnyCore/Backends/Java/JavaCodeGenerator.cs [hunk 3]
```diff
@@ -3606,6 +3606,10 @@ protected class ClassWriter : IClassWriter {
         message = "unexpected control point";
       }
 
+      // Wrapping an "if (true) { ... }" around the "break" statement is a way to tell the Java compiler not to give
+      // errors for any (unreachable) code that may follow.
```
Comment (atomb) [inline Source/DafnyCore/Backends/Java/JavaCodeGenerator.cs:3610]: :-P

**Review by atomb — DISMISSED** [review]
I think this looks good. There's an `#if` block that I wasn't sure you intended to leave in or not, but I'll let you decide whether or not to remove it.

File: Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.cs (hunk 2 above)
Comment (RustanLeino) [reply Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.cs:2754]: Oops. That was leftover from straightening out a merge. Thanks for catching that.

**Review by atomb — APPROVED** [review]

**Pull request merged**